### ISSUE TITLE
Improve performance

### ISSUE THREAD

**Issue by sindresorhus** [sindresorhus/globby#50]
This issue is dedicated to gathering ideas on how we can improve performance.

Some things that might help is to do filtering in the correct order to ensure minimal work is done, and cache and memoize as much as possible. The most important thing is that we skip ignored directories like `node_modules` as early as possible. It's the source of a *lot* of performance problems.

Some more ideas in #44 and mrmlnc/fast-glob#15

**Comment by zachleat** [comment]
FWIW, the biggest performance bottleneck in my testing on [eleventy](https://github.com/11ty/eleventy/) (for whatever reason) was the `gitignore: true` option. Removing this and just looping over `.gitignore` myself and adding the entries as `!` lines was much faster, for whatever reason.

(Love globby—thanks!)

**Comment by yaodingyd** [comment]
About filtering out `.gitignore` patterns before globbing, is translating gitignore rules and adding them to fast-glob ignore the right direction to go?

**Comment by sindresorhus** [comment]
> About filtering out `.gitignore` patterns before globbing, is translating gitignore rules and adding them to fast-glob ignore the right direction to go?

I think so, yes.

**Comment by sindresorhus** [comment]
> I'm adding this here as it's somewhat related: right now just requiring `globby` on my system often takes anywhere between 100ms and 250ms, and just installing it on a new directory downloads more than 4MB of stuff.

Yeah, that is quite unfortunate. It's because of `micromatch`, a dependency of `fast-glob`. See: https://github.com/mrmlnc/fast-glob/issues/154

<img width="924" alt="screen shot 2019-02-15 at 18 41 10" src="https://user-images.githubusercontent.com/170270/52854544-4d452580-3151-11e9-89d5-d5f8a62956ff.png">

// @mrmlnc

**Comment by dwelle** [comment]
I concur with @zachleat that the `gitignore: true` has a huge perf cost for whatever reason.

On our codebase:

1. using `gitignore: true`:

    ~8sec
	
    ```js
    await globby([ `**` ], { gitignore: true });
    ```

2. custom:

    ~100ms

    ```js
    const ignored = fs.readFileSync(".gitignore", { encoding: "utf8" })
        .split("\n")
        .filter( line => !/^\s*$/.test(line) && !/^\s*#/.test(line))
        .map( line => "!" + line.trim().replace(/^\/+|\/+$/g, ""));

    const files = await globby([
        `**`,
        `!.git`,
        ...ignored
    ]);
    ```

~4800% difference is a lot...

**Comment by janat08** [comment]
rollup is said to be good at removing dead code, I don't know if it goes as far as to repackage packages.

**Comment by kangax** [comment]
Just ran into this as well in our codebase. Commenting out gitignore leads to an improvement from 500 sec to 2 sec.

**Comment by DavidAnson** [comment]
For folks complaining about the performance hit of the `Options.gitignore`, my understanding of the current implementation is enabling that option adds a blocking preprocessing step that enumerates and parses potentially **every** `.gitignore` file in the directory tree regardless of the specified `patterns` value. It looks like `Options.ignore` is respected, but for many (perhaps most?) inputs to `globby`, the scope of this additional work may be **vastly** larger than what's done when `gitignore` is not specified.

Some of this additional work is unnecessary in some conditions, but I don't know offhand how difficult it is to determine that at runtime given the recursive nature of `.gitignore` and `patterns`.

Reference:
- https://github.com/sindresorhus/globby/blob/b0d7330942a3d36a12056f3a1257f676a534b6ea/index.js#L104
- https://github.com/sindresorhus/globby/blob/b0d7330942a3d36a12056f3a1257f676a534b6ea/ignore.js#L22
- https://github.com/sindresorhus/globby/blob/b0d7330942a3d36a12056f3a1257f676a534b6ea/ignore.js#L65

**Comment by DavidAnson** [comment]
> On our codebase:
> 1. using `gitignore: true`:
>    ~8sec
> 2. custom:
>    ~100ms
> ~8000% difference is a lot...

FYI, this is not an apples-to-apples comparison as the `gitignore` option for `globby` honors *every* `.gitignore` file in the tree while the custom code honors only the *one* at the root.

**Comment by DavidAnson** [comment]
For scenarios where only the root `.gitignore` file matters, setting `Options.ignoreFiles` to `".gitignore"` should be very fast as it should avoid the (correct by specification!) `**` part of the default behavior.

**Comment by fabiospampinato** [comment]
By coincidence I had written a little [benchmark](https://twitter.com/fabiospampinato/status/1774106797902594153) comparing globby against [`tiny-readdir-glob-gitignore`](https://github.com/fabiospampinato/tiny-readdir-glob-gitignore), here's the result:

![bench](https://github.com/sindresorhus/globby/assets/1812093/10706221-a8a9-4c20-87de-9b0164526ed4)

Basically as mentioned in the original post indeed skipping ignored directories as fast as possible is significant. Additionally a much faster implementation than `ignore` is possible. And in all when optimized enabling handling of `.gitignore`-like files should speed things up rather than slow things down in most real world scenarios, I think.

**Comment by danielbayley** [comment]
> By coincidence I have recently written …

@fabiospampinato Nice! I was just in the process of writing my own ideal glob function—based on `fs.readdir` and `picomatch`—before coming across `tiny-readdir-glob`[`-gitignore`]… So maybe now I don’t have to :)

But just have a couple of questions:

1. How does performance compare to _[fdir](https://github.com/thecodrr/fdir)_?
2. Why the need for [tiny-readdir](https://github.com/fabiospampinato/tiny-readdir), rather than `import {readdir} from 'node:fs/promises`?

**Comment by fabiospampinato** [comment]
@danielbayley could you open an issue about that in the other repo? That seems unrelated to `globby`.

**Comment by LastDragon-ru** [comment]
I our [case](https://github.com/DavidAnson/markdownlint-cli2/issues/365) enabling `gitignore` increase processing time from <1min to ~40min 😱
